psydac/linalg/iterative_solvers.py

```diff
@@ -521,3 +525,503 @@ def bicg(A, At, b, x0=None, tol=1e-6, maxiter=1000, verbose=False):
+        gamma  = max(gamma, eps)
+        cs     = gbar / gamma                # ck
+        sn     = beta / gamma                # sk
+        phi    = cs * phibar                 # phik
+        phibar = sn * phibar                 # phibark+1
+
+        # Update  x.
+
+        denom = 1.0/gamma
+        w1    = w2
+        w2    = w
+        w     = (v - oldeps*w1 - delta*w2) * denom
+        x     = x + phi*w
+
+        # Go round again.
+
+        gmax = max(gmax, gamma)
+        gmin = min(gmin, gamma)
+        z    = rhs1 / gamma
+        rhs1 = rhs2 - delta*z
+        rhs2 = - epsln*z
+
+        # Estimate various norms and test for convergence.
+
+        Anorm = sqrt(tnorm2)
+        ynorm = sqrt(x.dot(x))
+        epsa  = Anorm * eps
+        epsx  = Anorm * ynorm * eps
+        epsr  = Anorm * ynorm * tol
+        diag  = gbar
+
+        if diag == 0:diag = epsa
+
+        rnorm  = phibar
+        if ynorm == 0 or Anorm == 0:test1 = inf
+#        else:test1 = rnorm / (Anorm*ynorm)  # ||r||  / (||A|| ||x||)
+        else:test1 = rnorm                   # ||r||
+
+        if Anorm == 0:test2 = inf
+        else:test2 = root / Anorm           # ||Ar|| / (||A|| ||r||)
+
+        # Estimate  cond(A).
+        # In this version we look at the diagonals of  R  in the
+        # factorization of the lower Hessenberg matrix,  Q * H = R,
+        # where H is the tridiagonal matrix from Lanczos with one
+        # extra row, beta(k+1) e_k^T.
+
+        Acond = gmax/gmin
+
+        if verbose:
+            print( template.format(itn, rnorm ))
+
+        # See if any of the stopping criteria are satisfied.
+        if istop == 0:
+            t1 = 1 + test1      # These tests work if tol < eps
+            t2 = 1 + test2
+            if t2 <= 1:istop = 2
+            if t1 <= 1:istop = 1
+
+            if Acond >= 0.1/eps:istop = 4
+
+            if test2 <= tol:istop = 2
+            if test1 <= tol:istop = 1
+
+        if istop != 0:
+            break
+
+    if verbose:
+        print( "+---------+---------------------+")
+
+    # Convergence information
+    info = {'niter': itn, 'success': rnorm<tol, 'res_norm': rnorm }
+    return x, info
+# ...
+
+# ...
+def lsmr(A, At, b, x0=None, tol=None, atol=None, btol=None, maxiter=1000, conlim=1e8, verbose=False):
+    """Iterative solver for least-squares problems.
+    lsmr solves the system of linear equations ``Ax = b``. If the system
+    is inconsistent, it solves the least-squares problem ``min ||b - Ax||_2``.
+    ``A`` is a rectangular matrix of dimension m-by-n, where all cases are
+    allowed: m = n, m > n, or m < n. ``b`` is a vector of length m.
+    The matrix A may be dense or sparse (usually sparse).
+
+
+    Parameters
+    ----------
+    A : psydac.linalg.basic.LinearOperator
+        Left-hand-side matrix A of linear system; individual entries A[i,j]
+        can't be accessed, but A has 'shape' attribute and provides 'dot(p)'
+        function (i.e. matrix-vector product A*p).
```

Comment:
The `At` parameter is missing, see docstring of `bicg` function.

Reply:
fixed